src/mlpack/tests/ham_test.cpp

```diff
@@ -0,0 +1,280 @@
+  TreeMemory<double> mem(1, 1, J, W);
+  std::vector<std::vector<double>> initMemSTL = {{0}};
+  arma::mat initMem = convertToArma(initMemSTL);
+  mem.Initialize(initMem);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(0, 0), 0.);
+  mem.Update(0, arma::mat("12."));
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(0, 0), 12.);
+}
+
+BOOST_AUTO_TEST_CASE(TreeMemoryTestMinimumNDim)
+{
+  /*AddJoiner<double> J;
+  ReplaceWriter<double> W;*/
+  size_t memSize = 5;
+  FFN<MeanSquaredError<> > J = AddJoiner(memSize), W = ReplaceWriter(memSize);
+  TreeMemory<double> mem(
+      1, memSize, J, W);
+  std::vector<std::vector<double>> initMemSTL = {{0, 0, 0, 0, 0}};
+  arma::mat initMem = convertToArma(initMemSTL);
+  mem.Initialize(initMem);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, memSize);
+  for (size_t i = 0; i < memSize; ++i) {
+    BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(i, 0), 0.);
+  }
+  mem.Update(0, 12 * arma::ones(memSize, 1));
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, memSize);
+  for (size_t i = 0; i < memSize; ++i)
+    BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(i, 0), 12.);
+}
+
+BOOST_AUTO_TEST_CASE(TreeMemoryTestPowerOfTwo)
+{
+  /*AddJoiner<double> J;
+  ReplaceWriter<double> W;*/
+  FFN<MeanSquaredError<> > J = AddJoiner(1), W = ReplaceWriter(1);
+  TreeMemory<double> mem(8, 1, J, W);
+  std::vector<std::vector<double>> initMemSTL =
+      {{0}, {0}, {0}, {0}, {0}, {0}, {0}, {0}};
+  arma::mat initMem = convertToArma(initMemSTL);
+  mem.Initialize(initMem);
+  for (size_t idx = 0; idx < 15; ++idx)
+    BOOST_REQUIRE_EQUAL(mem.Cell(idx).at(0, 0), 0);
+  mem.Update(0, arma::mat("1."));
+  mem.Update(1, arma::mat("2."));
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).at(0, 0), 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(0).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Leaf(1).at(0, 0), 2);
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.LeafIndex(0))).at(0, 0), 3);
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.LeafIndex(0)))).at(0, 0), 3);  
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.Parent(mem.LeafIndex(0))))).at(0, 0), 3); 
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.LeafIndex(0))).n_elem, 1);
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.LeafIndex(0)))).n_elem, 1);  
+  BOOST_REQUIRE_EQUAL(mem.Cell(
+    mem.Parent(mem.Parent(mem.Parent(mem.LeafIndex(0))))).n_elem, 1); 
+}
+
+BOOST_AUTO_TEST_CASE(TreeMemoryTestPowerOfTwoNDim)
+{
+  /*AddJoiner<double> J;
+  ReplaceWriter<double> W;*/
+  size_t memSize = 4;
+  FFN<MeanSquaredError<> > J = AddJoiner(memSize), W = ReplaceWriter(memSize);
+  TreeMemory<double> mem(
+      8, memSize, J, W);
+  std::vector<std::vector<double>> initMemSTL =
+      {{0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0},
+       {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 0}};
+  arma::mat initMem = convertToArma(initMemSTL);
```

Comment:
Well, I did it for clarity (just like in every other test). In the "internal" `arma::mat` representation memory vectors are aligned column-wise, which imho is not human-intuitive. Besides, there is a convenient "standard" way to initialize `std::vector` (in `arma` it turns out to be platform-dependent). Taking these two together, I decided to initialize STL vector (intuitive) and then convert it to `arma` (hides complexity)